src/class.rs

```diff
@@ -6,6 +6,29 @@ const DFU_SUBCLASS_FIRMWARE_UPGRADE: u8 = 0x01;
 const DFU_PROTOCOL_RUNTIME: u8 = 0x01;
 const DFU_TYPE_FUNCTIONAL: u8 = 0x21;
 const DFU_REQ_DETACH: u8 = 0;
+const DFU_REQ_GETSTATUS: u8 = 0x03;
+const DFU_REQ_CLRSTATUS: u8 = 0x04;
+
+
+#[repr(u8)]
+#[derive(Clone, Copy)]
+enum DfuState {
+    /// Device is running its normal application.
+    AppIdle = 0,
+    /// Device is running its normal application, has received the DFU_DETACH request, and is waiting for a USB reset.
+    AppDetach = 1,
+    // An error ocurred, waiting for CLRSTATUS request.
+    DfuError = 10,
```

Comment:
I think we shouldn't return dfuERROR state when in runtime mode as it belongs to the DFU mode. In the state transition tables A.2.1/A.2.2 of DFU specification (rev 1.1) the "Next state" is always app* (besides switching to DFU mode), even after stalling the endpoint, e.g. `Receipt of any other DFU class-specific request` -> `Device stalls the control pipe.` -> `appIDLE`.

I tested this using `dfu-util` with my device that only allows DFU after pressing a button and it looks like `dfu-util` (or libusb?) has problems when it sees dfuERROR in runtime mode:
```
Claiming USB DFU (Run-Time) Interface...
Setting Alternate Interface zero...
Determining device status...
DFU state(10) = dfuERROR, status(15) = Device stalled an unexpected request
dfuERROR, clearing status
dfu-util: WARNING: Device already in DFU mode? (bState=10 dfuERROR)
Claiming USB DFU Interface...
Setting Alternate Interface #0 ...
Determining device status...
DFU state(0) = appIDLE, status(0) = No error condition is present
dfu-util: Device still in Run-Time Mode!
```
and it does not detach even though it should.

So it looks like this:
1. try to detach using `dfu-util`
1. fails because user did not allow
1. users presses "allow" button
1. try to detach using `dfu-util`
1. `dfu-util` fails with the above message, even though the firmware would detach now
1. try to detach using `dfu-util` (again)
1. device detaches

When we stop returning dfuERROR the steps 5 and 6 are not necessary.

This has the downside that without the `dfuERROR` state `dfu-util` doesn't seem to make CLRSTATUS request, so the status is never cleared until the device detaches to DFU mode, but this doesn't seem to cause any actual problems, just that the errSTALLEDPKT is reported.

What do you think about this apprach? Could you test it with https://fwupd.org?